File: library/src/main/java/renetik/android/view/CSPagerAdapter.java
package renetik.android.view;

import android.view.View;
import android.view.ViewGroup;

import androidx.viewpager.widget.PagerAdapter;
import renetik.android.java.collections.CSList;
import renetik.android.viewbase.CSViewController;

import static renetik.android.java.collections.CSListKt.list;

public class CSPagerAdapter<T extends CSViewController & CSPagerPage> extends PagerAdapter {

    private CSList<T> _controllers = list();

    public CSPagerAdapter(CSList<T> controllers) {
        _controllers = controllers;
    }

    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
    }

    public int getCount() {
        return _controllers.count();
    }

    public CharSequence getPageTitle(int position) {
        return _controllers.at(position).csPagerControllerTitle();
    }

    public Object instantiateItem(ViewGroup container, int position) {
        View view = _controllers.at(position).getView();
        container.addView(view, 0);
        return view;
    }

    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

}

File: library/controller/src/main/java/pager/CSPagerAdapter.kt
package renetik.android.controller.pager

import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import renetik.android.controller.base.CSViewController
import renetik.android.java.collections.CSList
import renetik.android.java.collections.list

class CSPagerAdapter<PagerPageType>(val controllers: CSList<PagerPageType> = list<PagerPageType>())
    : PagerAdapter() where PagerPageType : CSViewController<*>, PagerPageType : CSPagerPage {

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) =
            container.removeView(`object` as View)

    override fun getCount() = controllers.size

    override fun getPageTitle(position: Int) = controllers[position].pagerPageTitle

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val view = controllers[position].view
        container.addView(view, 0)
        return view
    }

    override fun isViewFromObject(view: View, `object`: Any) = view === `object`

}
